internal/client/db/db.go

```diff
@@ -0,0 +1,585 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package db
+
+import (
+	"fmt"
+	"log"
+	"sync"
+
+	"github.com/ChainSafe/gossamer/internal/client/api"
+	"github.com/ChainSafe/gossamer/internal/client/db/columns"
+	"github.com/ChainSafe/gossamer/internal/client/db/metakeys"
+	"github.com/ChainSafe/gossamer/internal/primitives/blockchain"
+	"github.com/ChainSafe/gossamer/internal/primitives/core/hash"
+	"github.com/ChainSafe/gossamer/internal/primitives/database"
+	"github.com/ChainSafe/gossamer/internal/primitives/runtime"
+	"github.com/ChainSafe/gossamer/internal/primitives/runtime/generic"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+	"github.com/li1234yun/gods-generic/maps/linkedhashmap"
+)
+
+const numCachedHeaders = 8
+
+// An extrinsic entry in the database.
+type dbExtrinsic[E runtime.Extrinsic] struct {
```

Comment:
given that those database entries are just handled at host level (don't have any need to be passed to the runtime) why not create their own struct instead of implement them as enums? 

With the following suggestion you can have a less complex code and still use the `switch .(type)` approach, you can even scale encode each type as a struct instead the overhead of encoding a SCALE enum

```go
type DbExtrinsic interface {
  isExtrisic()
}

type ExtrinsicIndexed struct {
         Hash hash.H256
	// Extrinsic header.
        Header []byte
}
func (ExtrinsicIndexed) isExtrisic() {}

type ExtrinsicFull struct {
          Extrinsic runtime.Extrinsic <- can be defined as an interface instead of a generic
}
func (ExtrinsicFull) isExtrisic() {}

```

Reply:
I implemented it as a `VaryingDataType` since this is how it's encoded and stored in the underlying db.  I would need a way to differentiate between the indexed vs full extrinsic when decoding from the underlying db if I were to move away from SCALE.  I can revisit this later once I validate that methods like `BlockIndexedBody` are returning the right data, but I think it's a prudent approach right now to try and replicate the same way substrate is storing these `DbExtrinsic` given I haven't implemented the broader `db.Backend` yet.